lib/connections/service.go

```diff
@@ -669,8 +708,17 @@ func (s *service) resolveDialTargets(ctx context.Context, now time.Time, cfg con
 
 		dialer := dialerFactory.New(s.cfg.Options(), s.tlsCfg, s.registry, s.lanChecker)
 		priority := dialer.Priority(uri.Host)
-		if priority >= priorityCutoff {
-			l.Debugf("Not dialing using %s as priority is not better than current connection (%d >= %d)", dialerFactory, priority, priorityCutoff)
+		currentConns := s.numConnectionsForDevice(deviceCfg.DeviceID)
+		if priority >= priorityCutoff && currentConns >= s.desiredConnectionsToDevice(deviceCfg.DeviceID) {
+			l.Debugf("Not dialing %s at %s using %s as priority is not better than current connection (%d >= %d) and we already have %d/%d connections", deviceID.Short(), addr, dialerFactory, priority, priorityCutoff, currentConns, deviceCfg.NumConnections)
+			continue
+		}
+		if priority > priorityCutoff {
+			l.Debugf("Not dialing %s at %s using %s as priority is worse than current connection (%d > %d)", deviceID.Short(), addr, dialerFactory, priority, priorityCutoff)
+			continue
+		}
```

Comment:
The order with overlapping priority conditions had me confused for a second - might be nicer to change the order of conditions and use `==` then.

Reply:
Clarified!